discovery/config.go

```diff
@@ -47,9 +53,26 @@ func (c Config) withDefaults() Config {
 	if c.ServerWatchDisabledInterval == 0 {
 		c.ServerWatchDisabledInterval = DefaultServerWatchDisabledInterval
 	}
+
+	// Infer the ServerName field if a hostname is used in Addresses.
+	if c.TLS != nil && !c.TLS.InsecureSkipVerify && c.TLS.ServerName == "" && isPotentialHostname(c.Addresses) {
+		c.TLS = c.TLS.Clone()
+		c.TLS.ServerName = c.Addresses
+	}
+
 	return c
 }
 
+// isPotentialHostname returns true if the addr is not an IP or "exec=" string,
+// and true otherwise. It may return true for something that is not actually a
+// valid hostname.
+func isPotentialHostname(addr string) bool {
+	if addr == "" || strings.HasPrefix(addr, "exec=") {
+		return false
+	}
+	return net.ParseIP(addr) == nil
```

Comment:
I don't think we accept an IP for the addresses field but I guess it doesn't hurt to check!

Reply:
Oh, go-netaddrs does accept a single IP address, but maybe that is unintentional 🤔

I think we could skip the IP check here to simplify. Then it might set the ServerName to an IP address if `Addresses = "<ip>", which is unusual...but I have used `127.0.0.1` for the SAN field in local testing with a self-signed cert so there's that 🤷

Reply:
I removed the IP check here https://github.com/hashicorp/consul-server-connection-manager/pull/6/commits/32ea6307b58e2184df8bb0d054fcd0d8a0d9ef04

It now only checks that the `Addresses` field is not a command (does not start with `exec=`).